rfcs/0000-kuadrant-release-process.md

```diff
@@ -0,0 +1,207 @@
+# Kuadrant Release Process
+
+- Feature Name: `kuadrant-release-process`
+- Start Date: 2024-01-11
+- RFC PR: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/pull/0000)
+- Issue tracking: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Kuadrant is a set of components that their artifacts are built and delivered independently. This RFC aims to define every
```

Comment:
I think that the difference is Kuadrant operator (the orchestrator of the suite) and Kuadrant components (Authorino, Limitador, MGC, etc). I'm saying this because I thought that QA mainly checks the versions included in the latest released version of Kuadrant Operator. I could re write it as per each component... which kind of will remove the responsibility of deciding the cadence and leave it to Pauls RFC (?)

Reply:
On another note, it mentions that the artifacts from each component are build (and delivered) independently, but might not be subject to a new event of releasing.

Reply:
> actual release process should be the same for every component

I see this differently.

IMO, it does not _have to_ be, neither it _can_ always be the same process for every component. Rather (as originally phrased), the components are independent from the whole and can be released independently, following their own particular release processes, respecting each one its own specificity. It's not top-down; it's bottom-up.

For some component, it may make sense, for example, doing release branches, while for others it's not needed and just a tag will do. Some components may involve publishing artifacts to a particular repo, while others do not. Because of these and so many other differences, each component is a project and has its own history. Thinking the opposite risks creating coupling between the components and generates scalability issues to the organisation.

Of course, we don't want multiple ways to do the exact same thing. To reduce overhead, if 2 components share a common step within their corresponding release processes, we should try performing that step in the same way – by resusing tools, automation, docs, etc across components. (Although, I think this would be something that belongs to #41 and not to this PR.)

Correct if I'm wrong, but I understand this RFC is for how to lock a version of what we call "Kuadrant", which is not the sum of its parts, but has an identity of its own. Therefore, it's not to force components all to stick to a same process but to highlight precisely this difference between the sum of the parts and the whole. Maybe it's a description of the steps to follow _after_ all components have been released, and perhaps some general guidelines that all components need to comply with, otherwise releasing Kuadrant would be impossible.

If there's a components that confuses itself with the whole, then an explanation of that and some words about how releasing that one component wraps up a release of Kuadrant itself maybe (if that's the case.)

Reply:
I would say this RFC should be high-level and only mention the release process, not the actual steps, e.g. how we build those artefacts should not be part of this (That is part of the Kuadrant CI/CD RFC and that indeed differs). I do not see a difference between component release and Kuadrant release in a high-level process. For each component there should be an RC build, and for each component QA should happen, exactly the way described here. The doc is also present for all components on the `kuadrant.io`, communication channels are also the same

So basically in my eyes, the process is the key section and it should apply to all component releases and the entire Kuadrant release at the same time.

Reply:
@pehala, I think you may be seeing Kuadrant == MGC (or any other components that currently sits at the root of the tree of dependencies.)

In case my view wasn't clear... IMO, Kuadrant is a composition of things (Kuadrant Operator, Authorino, Limitador, MGC) and a way to use those things. Let's call this second part "other stuff" – docs, templates, maybe an installer, and, in general, other files we can/could provide in the "Kuadrant package" that are not part of the none of the individual components per se.

So:

```
K = KOp + A + L + MGC + stuff
```

A given version `K_v1` of Kuadrant can be, for example, a composition of Kuadrant Operator v0.3, Authorino v15, Limitador v0.8 and MGC v0.1. I.e.:

```
K_v1 = KOp_v0.3 + A_v15 + L_v0.8 + MGC_v0.1 + stuff
```

If we want to bump the version of, say, Limitador that is used in/by Kuadrant to v0.9 - not because Limitador v0.9 has just been released; maybe the latest Limitador is already v2.5 by the time this happens –, then the Kuadrant composition will change. IOW, Kuadrant will be a different thing than what it was at v1. Maybe we are talking about Kuadrant v2:

```
K_v2 = KOp_v0.3 + A_v15 + L_v0.9 + MGC_v0.1 + stuff
```

The only thing that changed was Limitador and, maybe, some of the other `stuff`, like new docs saying that now Kudrant v2 comes with Limitador v0.9.

Each component, as well as the suite itself, has (should have) its own set of artifacts, its own version number, its own change log and release notes, its own testing and release process. Kuadrant is no longer equivalent to the Kuadrant Operator. MGC may have taken its place as the new root, but we need to see (...and test, ...and version, ...and release) Kuadrant independently from it as well.

This is the way I see Kuadrant, at least. I may change my mind if most people agree that Kuadrant == MGC, in which case I would argue for changes beyond the defining our release process, starting by renaming those repos, MGC and Kuadrant Operator, and moving some code and responsibilities around. In the end, it would all contribute to defining what releasing Kuadrant means and how it should look like.

Reply:
I kind of like the idea you bring defining what would be the version of _Kuadrant_, quoting your formula:

> K = KOp + A + L + MGC + stuff

In that case, and if everyone/most people agree, I could add another section regarding the particular versioning and grouping that will define it.

Reply:
I am all for the section defining Kuadrant and its related components. For the issue of the individual component release process, my suggestion would be to use this process for operator-based components that are released individually, meaning, this release process would be used for these components:

- Kuadrant as a whole
- Authorino + Authorino operator
- Limitador + Limitador Operator
- DNSOperator
- Multicluster Gateway Controller

These components are all operator-based and distributed the same way; thus, it makes sense for them to follow the same process.

@guicassolato @didierofrivia Is that ok with you? Does that even make sense?

Reply:
@pehala thanks for you comment! Let me elaborate and see if I understood correctly.... within a release of Kuadrant, one will encounter Operators released independently, which I reckon all of us agree (at least to a certain point) that given their similarities its easy to understand they will follow the same technical process of releasing... 

But given your last comment and trying to summarize the thread,  are we meaning to do...

```
Kuadrant vx.y.z = Kuadrant Operator vx.y.z + Authorino Operator vx.y.z + Limitador Operator vx.y.z + DNS Operator vx.y.z + MGC Controller vx.y.z
```
Where each dependency component will be released independently and grouped under a version of Kuadrant and then we could assign the Kuadrant version in the same fashion of the other components.

OR

Keeping the same way of independently releasing but matching the version of Kuadrant to its Kuadrant Operator:

```
Kuadrant vx.y.z = Kuadrant Operator vx.y.z =  Authorino Operator vx.y.z + Limitador Operator vx.y.z + DNS Operator vx.y.z + MGC Controller vx.y.z
```

Maybe I'm missing the point, but that is what we still haven't agreed on. 

@guicassolato @pehala how close are we to (dis)agree on this? Thanks in advance!

Reply:
When it comes to this, I have no preference, so whatever works for you

Reply:
```
Kuadrant vX.Y.Z = Kuadrant Operator vX.Y.Z =  Authorino Operator vA.B.C + Limitador Operator vD.E.F + DNS Operator vG.H.I + MGC Controller vJ.K.L
```